### PR TITLE
fix: improve swapper loading state

### PR DESCRIPTION
**Pull request by 0xApotheosis**
## Description

This PR improves loading states when loading and initialising the swapper (notably for KeepKey users, for which the initialisation is particularly slow during communication with the hardware).

We add a value to determine if we are currently initialising the swapper, and if so, ignore validation errors by early-exiting on the "loading" state.

## Notice

- [x] Have you followed the guidelines in our [Contributing]('https://github.com/shapeshift/web/blob/develop/CONTRIBUTING.md) guide?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/shapeshift/web/pulls) for the same update/change?

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [x] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

Closes https://github.com/shapeshift/web/issues/3797

## Risk

Small.

## Testing

We should no longer render "Invalid asset pair" on first load with a KeepKey.

### Engineering

☝️

### Operations

☝️

## Screenshots (if applicable)

Screenshot 2023-04-04 at 2 39 16 pm
Screenshot 2023-04-04 at 2 39 37 pm
Screenshot 2023-04-04 at 2 39 46 pm

### CODE REVIEW

File: src/components/Trade/TradeInput.tsx [hunk 1]
```diff
@@ -78,6 +78,7 @@ export const TradeInput = () => {
   useSwapperService()
   const [isLoading, setIsLoading] = useState(false)
   const [showQuotes, toggleShowQuotes] = useToggle(false)
+  const [isInitialLoad, setIsInitialLoad] = useState(true)
```
Comment (gomesalexandre) [inline src/components/Trade/TradeInput.tsx:81]: This is quite high risk - from the semantics of it, it looks like this is supposed to be true on the first load (mount) of any `<TradeInput />` instance, but this will be set to true on any mount of any `<TradeInput />` component instance (note the double *any*).

I have two concerns with this:

1. The naming, which seem to indicate intent of a loading state for *any* instance, whereas what this actually does is being set to true on the first *mount* of a *specific* instance
2. The `useEffect` below, which will only `setInitialLoad(false)` if `isSwapperApiPending`, which may never be the case during the lifecycle of a specific instance

tl;dr agree with the semantics, but the actual flow isn't matching the intent

File: src/components/Trade/TradeInput.tsx [hunk 2]
```diff
@@ -198,6 +199,14 @@ export const TradeInput = () => {
     [updateAction, updateIsSendMax, updateAmount, handleInputAmountChange],
   )
 
+  /*
+     On initial load we are missing a swapper, among other things, which causes invalid errors and warning to be shown
+     to the user. We want to wait until we've started making network calls before validating the state.
+   */
```
Comment (gomesalexandre) [inline src/components/Trade/TradeInput.tsx:205]: I also have concerns about these heuristics, regardless of whether the initial load state lives in a singleton or in a reactive instance (e.g a hook/a component or whatever might have different instances):

It looks like the intent is really to detect that we've started making network calls, e.g there's any network request, either pending, fulfilled, or rejected. `uninitialized` should work too, since it's only used in case of a `{skip: true}` RTK query hooks option which we don't seem to use anywhere across the `swapperApi` domain. In other words, if `swapperApi.queries` has any member, then we've started making network calls.

In this case, would it make sense to have a new `swapperApi.queries` selector to achieve the same effect as this e.g `Object.keys(state.swapperApi.queries).length`? It would be 0-length on the *first* load of the *first* swapper instance, and have a truthy length as soon as said instance first mounts and start making requests.

**Review by gomesalexandre — CHANGES_REQUESTED** [review]
Tested on KK, confirmed this fixes the loading state and errors on initial load, though brings regressions across all wallets (tested with KK, native, and MM):

#### When going to the trade page, waiting a few seconds for the swapperApi queries to fullfill then leaving the trade page and coming back, the swapper stays in an infinite loading state

image

- Looks like the intent here is for this to happen on the first load of any swapper only? i.e coming back should show a grayed-out "Preview Trade"?

#### Similarly, when getting a trade for a specific pair / amount (in this case the default ETH -> FOX), leaving the swapper, then coming back and setting the same values as previously (i.e same pair/amount), the swapper is in an infinite loading state:

- First time inputting any arbitrary amount

image

- Second time inputting the same amount after leaving the trade page

image

In both cases noted above, clicking the "Loading" button seemingly does nothing as a user (though it does fire XHRs)

Inputting a new amount or selecting a new pair does fix this state, since it results in new swapper APIs to be pending, thus the effect running.

**Comment by 0xApotheosis** [comment]
You're absolutely right @gomesalexandre, the previous heuristics were poorly considered.

Updated in https://github.com/shapeshift/web/pull/4185/commits/87d72ac02ab803ff829502ac955e7843f8d729ee to use queries length, which better represents the intend of the what were are trying to achieve here.

**Review by gomesalexandre — APPROVED** [review]
Tested locally, confirmed the bugs spotted previously on all wallets are now gone with the consumption of `selectSwapperQueriesInitiated` 🎉 

Also confirmed KK is now happy while swapper queries are not yet initiated:

image
image
